pallets/corporate-actions/src/ballot.rs

```diff
@@ -0,0 +1,544 @@
+        ///    The full voting power of `origin`'s DID may be used for each motion in the ballot.
+        ///
+        /// # Errors
+        /// - `NoSuchBallot` if `ca_id` does not identify a ballot.
+        /// - `VotingNotStarted` if the voting period hasn't commenced yet.
+        /// - `VotingAlreadyEnded` if the voting period has ended.
+        /// - `WrongVoteCount` if the number of choices in the ballot does not match `votes.len()`.
+        /// - `NoSuchCA` if `ca_id` does not identify an existing CA.
+        /// - `NotTargetedByCA` if the ballot does not target `origin`'s DID.
+        /// - `InsufficientVotes` if the voting power used for any motion in `votes`
+        ///    exceeds `origin`'s DID's voting power.
+        #[weight = 950_000_000]
+        pub fn vote(origin, ca_id: CAId, votes: Vec<(T::Balance, u16)>) {
+            let did = <Identity<T>>::ensure_perms(origin)?;
+
+            // Ensure ballot has started but not ended, i.e. `start <= now <= end`.
+            let range = Self::ensure_ballot_exists(ca_id)?;
+            let now = <Checkpoint<T>>::now_unix();
+            ensure!(range.start <= now, Error::<T>::VotingNotStarted);
+            ensure!(now <= range.end, Error::<T>::VotingAlreadyEnded);
+
+            // Ensure we have balances provided for each choice.
+            let choices_count = MotionNumChoices::get(ca_id);
+            let total_choices = choices_count.iter().copied().map(|c| c as usize).sum::<usize>();
+            ensure!(votes.len() == total_choices, Error::<T>::WrongVoteCount);
+
+            // Divide `votes` into motions.
+            let motions = choices_count
+                .iter()
+                .map(|c| *c as usize)
+                .scan(0, |start, count| Some(&votes[mem::replace(start, *start + count)..count]));
+
+            if RCV::get(ca_id) {
+                // RCV is enabled.
+                // Ensure that all fallback choices point to some choice in the same motion.
+                motions
+                    .clone()
+                    .all(|votes| {
+                        let count = votes.len();
+                        votes
+                            .iter()
+                            .map(|vote| vote.1 as usize)
+                            .enumerate()
+                            .all(|(idx, fb)| idx != fb && fb < count)
+                    })
+                    .then_some(())
+                    .ok_or(Error::<T>::NoSuchRCVFallback)?;
+            } else {
+                // It's not. Make sure its also not used.
+                votes
+                    .iter()
+                    .all(|(_, fb)| *fb == 0)
+                    .then_some(())
+                    .ok_or(Error::<T>::RCVNotAllowed)?;
+            }
+
+            // Ensure that `did` is targeted by this ballot.
+            let ca = <CA<T>>::ensure_ca_exists(ca_id)?;
+            ensure!(ca.targets.targets(&did), Error::<T>::NotTargetedByCA);
```

Comment:
Strictly speaking, if `ensure_ballot_exists` is successful, then `ensure_ca_exists` cannot fail, but better safe than sorry.

As for "initial", do you mean that we should move the targets check earlier? If so, then I tend to agree, since its an early cheap `O(log n)` check.

Reply:
Yeah, I mean to move it early.